### PR TITLE
Add type hints to folder_paths.py

### PR DESCRIPTION
**Pull request by PhilWun**
This PR adds type hints to function declarations and some variables in folder_paths.py to improve the developer experience. If its OK I would add type hints to more parts of the codebase in future PRs.

### CODE REVIEW

**Review by huchenlei — APPROVED** [review]
LGTM! Thanks for the effort! I think @Kosinkadink would really appreciate your work here

File: folder_paths.py [hunk 1]
```diff
@@ -42,41 +43,41 @@
 input_directory = os.path.join(os.path.dirname(os.path.realpath(__file__)), "input")
 user_directory = os.path.join(os.path.dirname(os.path.realpath(__file__)), "user")
 
-filename_list_cache = {}
+filename_list_cache: Dict[str, Tuple[List[str], Dict[str, float], float]] = {}
```
Comment (huchenlei) [inline folder_paths.py]: Minor: we can `from __future__ import annotations` to use the Python 3.10 syntax here, i.e. `dict[str, tuple[list[str], dict[str, float], float]]`

Comment (PhilWun) [reply folder_paths.py]: Thank you for your feedback. I implemented your suggestion.

**Comment by huchenlei** [comment]
@PhilWun Can you rebase to include https://github.com/comfyanonymous/ComfyUI/pull/4234? @comfyanonymous would like to see the typing is compatible with Python 3.8

File: folder_paths.py [hunk 2]
```diff
@@ -1,13 +1,15 @@
...
-from typing import Set, List, Dict, Tuple
+from collections.abc import Collection
 
-supported_pt_extensions: Set[str] = set(['.ckpt', '.pt', '.bin', '.pth', '.safetensors', '.pkl', '.sft'])
+supported_pt_extensions: set[str] = {'.ckpt', '.pt', '.bin', '.pth', '.safetensors', '.pkl', '.sft'}
 
-SupportedFileExtensionsType = Set[str]
-ScanPathType = List[str]
-folder_names_and_paths: Dict[str, Tuple[ScanPathType, SupportedFileExtensionsType]] = {}
+SupportedFileExtensionsType = set[str]
```
Comment (huchenlei) [inline folder_paths.py]: I think we should inline these 2 variables. My bad.
```
Traceback (most recent call last):
  File "main.py", line 6, in <module>
    import folder_paths
  File "/home/runner/work/ComfyUI/ComfyUI/ComfyUI/folder_paths.py", line 10, in <module>
    SupportedFileExtensionsType = set[str]
TypeError: 'type' object is not subscriptable
```

**Comment by PhilWun** [comment]
The test failed because even with `from __future__ import annotations`  subscripted buildin types like `set[str]` cannot be assigned to a variable in Python 3.8. I have found three possible solutions.

Current code:
```python
SupportedFileExtensionsType = set[str]
ScanPathType = list[str]
folder_names_and_paths: dict[str, tuple[ScanPathType, SupportedFileExtensionsType]] = {}
```
Solution 1.:
```python
from typing import List, Set

SupportedFileExtensionsType = Set[str]
ScanPathType = List[str]

folder_names_and_paths: dict[str, tuple[ScanPathType, SupportedFileExtensionsType]] = {}
```
Solution 2.:
```python
import sys

if sys.version_info < (3, 9):
    from typing import List, Set
    SupportedFileExtensionsType = Set[str]
    ScanPathType = List[str]
else:
    SupportedFileExtensionsType = set[str]
    ScanPathType = list[str]

folder_names_and_paths: dict[str, tuple[ScanPathType, SupportedFileExtensionsType]] = {}
```
Solution 3:
```python
folder_names_and_paths: dict[str, tuple[list[str], set[str]]] = {}
```

**Comment by huchenlei** [comment]
@comfyanonymous Gentle ping on merge.

**Comment by JettHu** [comment]
Writing union types as `X | Y` is [PEP 604](https://www.python.org/dev/peps/pep-0604), supported after python 3.10 only.

Should we switch back to the old method to improve compatibility, or add the Python version description in the README?

**Comment by melMass** [comment]
Not a big deal IMO but you keep trying to be compatible with 3.8, `| None` will error out there

**Pull request merged**